### PR TITLE
Add support for building on RHEL

### PR DESCRIPTION
**Pull request by omajid**
This adds support for building roslyn on RHEL using:

    $ make bootstrap RUNTIME_ID=rhel.7-x64
    $ make all RUNTIME_ID=rhel.7-x64

Please note that the `Makefile` uses `/etc/os-release` to compute the runtime id. Passing `RUNTIME_ID` on the command line (not as environment variable!) overrides that.

### CODE REVIEW

**Comment by ellismg** [comment]
@agocke @jaredpar 

I wonder if there is some way that we can not have to explicitly list every RID.  Would it be possible to do something like:

```
<RuntimeIdentifiers Conditon="'$(HostRID)' != ''">$(RuntimeIdentifiers);$(HostRID)</RuntimeIdentifiers>
```

Then build.sh or something could pass /p:HostRID=rhel.7-x64, and if you wanted to build on fedora or whatever, you could use the RID for that platform?

File: build/Toolset/Toolset.csproj [hunk 1]
```diff
@@ -10,7 +10,7 @@
     <RootNamespace>Toolset</RootNamespace>
     <AssemblyName>Toolset_DoNetUse</AssemblyName>
     <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
-    <RuntimeIdentifiers>win7;ubuntu.14.04;osx.10.10</RuntimeIdentifiers>
+    <RuntimeIdentifiers>win7;rhel.7;ubuntu.14.04;osx.10.10</RuntimeIdentifiers>
```
Comment (ellismg) [inline build/Toolset/Toolset.csproj]: There are a few places where we don't include an architecture when specifying a RID.  What's the reason behind that @jaredpar or @agocke?

**Comment by jaredpar** [comment]
> I wonder if there is some way that we can not have to explicitly list every RID.

That would be amazing.  Every time I ask around though I can't get a good way to do this.  Really wish MSBuild would have a concept of `$(DefaultRuntimeIdentifier)` that just makes to whatever MSBuild is presently running on.  It would make our build infrastructure so much simpler.  

Lately I have been trying to use our Makefile as a choke point for calculating the runtitme idintifier and passing it down as a build option.  See here as an example

https://github.com/dotnet/roslyn/blob/master/Makefile#L15

One overall comment about this PR.  If we take this change, supporting redhat in our build, I want to add CI validation that we don't break it.  Otherwise I feel like we'd break this a lot (as we do most of our Linux work on ubuntu).  What do we need to do to get this into our CI?  Is there a Redhat image we can use to constantly validate this build?

**Comment by ellismg** [comment]
> What do we need to do to get this into our CI? Is there a Redhat image we can use to constantly validate this build?

Yes, we already have RHEL images.  It should be as easy as looking at corefx's netci.groovy, which already has RHEL runs.  Do you want to take a look at that @omajid?

**Comment by ellismg** [comment]
> Please note that the Makefile uses /etc/os-release to compute the runtime id. Passing RUNTIME_ID on the command line (not as environment variable!) overrides that.

I know this isn't your doing, @omajid, but I wish there was a way for us just ask dotnet what it thought the RID for the current platform was instead of having to reproduce the logic here.  `dotnet --info` plus some grep'ing and sed'ing would let us do what we wanted, but we'd have to change the structure of the Makefile such that dotnet was already installed before we wanted to compute RUNTIME_ID.  Any thoughts @omajid @agocke @jaredpar

**Comment by ellismg** [comment]
@omajid I took a stab at an alternate approach: https://github.com/ellismg/roslyn/tree/build-rhel-7.3

With the above, you could do something like: `DOTNET_RUNTIME_ID=rhel.7-x64 ./cibuild.sh --skipcommitprinting` and I believe it should work.

`DOTNET_RUNTIME_ID` is understood by the tooling as a way to override the RID of the current platform, so maybe it makes sense to use that as a way to prime `RUNTIME_ID` and then fall back to /etc/os-release if it's not set (which is close to what the tooling would do, modulo fixups)

**Comment by agocke** [comment]
@ellismg I like this general approach. I'd prefer if we could just say "we support whatever CoreCLR supports" since we don't have any particular runtime dependencies aside from CoreCLR and the associated framework packages.

**Comment by omajid** [comment]
@ellismg that's a *much* better approach. I will gladly take that!

**Comment by omajid** [comment]
@ellismg I opened https://github.com/dotnet/roslyn/pull/19519 for CI

**Comment by omajid** [comment]
Closing this issue since the fix in https://github.com/dotnet/roslyn/pull/19560 is *much* better.

**Pull request closed without merging**